spec.md

```diff
@@ -1598,6 +1652,77 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 | Invalid `starting_token` | 10 ABORTED | Indicates that `starting_token` is not valid. | Caller SHOULD start the `ListSnapshots` operation again with an empty `starting_token`. |
 
 
+#### `ControllerExpandVolume`
+
+A Controller plugin MUST implement this RPC call if plugin has `EXPAND_VOLUME` controller capability.
+This RPC allows the CO to expand the size of a volume.
+
+This call MAY be made by the CO during any time in the lifecycle of the volume after creation if plugin has `VolumeExpansion.ONLINE` capability.
+If plugin has `EXPAND_VOLUME` node capability, then `NodeExpandVolume` MUST be called after successful `ControllerExpandVolume` and `fs_resize_required` in `ControllerExpandVolumeResponse` is `true`.
+
+If the plugin has only `VolumeExpansion.OFFLINE` expansion capability and volume is currently published or available on a node - `ControllerExpandVolume` MUST be called ONLY after either:
+- The plugin has controller `PUBLISH_UNPUBLISH_VOLUME` capability and `ControllerUnpublishVolume` has been invoked successfully.
+
+OR ELSE
+
+- The plugin does NOT have controller `PUBLISH_UNPUBLISH_VOLUME` capability, the plugin has node `STAGE_UNSTAGE_VOLUME` capability, and `NodeUnstageVolume` has been completed successfully.
+
+OR ELSE
+
+- The plugin does NOT have controller `PUBLISH_UNPUBLISH_VOLUME` capability, nor node `STAGE_UNSTAGE_VOLUME` capability, and `NodeUnpublishVolume` has completed successfully.
+
+Examples:
+* Offline Volume Expansion:
+  Given an ElasticSearch process that runs on Azure Disk and needs more space.
+  - The administrator takes the Elasticsearch server offline by stopping the workload and CO calls `ControllerUnpublishVolume`.
+  - The administrator requests more space for the volume from CO.
+  - The CO in turn first makes `ControllerExpandVolume` RPC call which results in requesting more space from Azure cloudprovider for volume-id that was being used by ElasticSearch.
+  - Once `ControllerExpandVolume` is completed and successful, the CO will inform administrator about it and administrator will resume the ElasticSearch workload.
+  - The node on which ElasticSearch workload is scheduled the CO calls `NodeExpandVolume` after calling `NodeStageVolume`.
```

Comment:
IIRC the conclusion of the last meeting was to make `NodeExpandVolume` exclusive for ONLINE expansion, right? If so, then this line and the one below are incorrect and should be removed. Instead we could mention that the plugin may choose to resize the filesystem in the next `NodeStageVolume` or `NodePublishVolume` call.

Reply:
Please correct me if I'm wrong. Maybe my understanding is outdated.

But the problem of making `NodeExpandVolume` involved in OFFLINE expansion is that the CO does not know when to safely call `NodeExpandVolume`. IIUC there are 3 types of filesystems in terms of volume expansion support:
1. The filesystem supports volume expansion no matter volumes are mounted or not.
2. The filesystem supports volume expansion only when the volumes is mounted (i.e., `VOL_READY` or `PUBLISHED`).
3. The filesystem supports volume expansion only when the volume is not mounted (i.e., `NODE_READY`).
Do you mean that since most modern filesystems are either of type 1 or 2, we just let `NodeExpandVolume` support these filesystems, and for type 3 the plugin has to do the filesystem resize in either `NodeStageVolume` or `NodePublishVolume` and leave `NodeExpandVolume` a no-op?

Reply:
Let me try to clarify the pros and cons of making `NodeExpandVolume` involved in an OFFLINE expansion, and compare it with ONLINE expansion.

In an ONLINE expansion, it is straight-forward for the CO to call `NodeExpandVolume` since the volume is already node-staged or node-published, and the CO knows where node is, so it can simply call `NodeExpandVolume` on that node.

However, in an OFFLINE expansion, it is more challenging, because the whole expansion is split into two steps (`ControllerExpandVolume` and `NodeExpandVolume`), and the CO needs to either 1. choose a valid node to call `NodeExpandVolume`, which involves topology checking; or 2. defer `NodeExpandVolume` until a workload consuming the volume is placed, which mean s that the CO needs to keep track of an additional state -- whether `NodeExpandVolume` has been called or not -- and change the state after the first workload is placed.

Instead, if `NodeExpandVolume` is just excluded in an OFFLINE expansion, it would be way easier for the CO to maintain the volume expansion state: in the CO's view, it's done once `ControllerExpandVolume` is done. WDYT?

Reply:
And if we do agree that, a volume plugin indeed can have both ONLINE and OFFLINE capabilities then how is that different from a plugin that has ONLY OFFLINE but has both `ControllerExpandVolume` and `NodeExpandVolume` ? Same scheduling problem you mentioned applies to both of them.

Reply:
Yeah I agree that we should not limit the plugin to only support either ONLINE or OFFLINE expansion. I also get that you have already make it clear in the spec about when to call `NodeExpandVolume` (Thanks for your efforts on this!). I'm not stating the "safety" concern precisely so please let me do that again: since CO can only call `NodeExpandVolume` when a volume is node-staged/node-published, for filesystems of type 3, the plugin cannot put the expansion logic in `NodeExpandVolume`.

My point is that, it would be way easier for COs if an OFFLINE expansion is considered "done" after `ControllerExpandVolume` and leave the plugin to do whatever necessary for filesystem changes on its own whenever appropriate. This is similar to how a plugin would do a `mkfs` right now -- at the first `NodeStageVolume` or `NodePublishVolume`, without any CO intervention. So this seems kinda consistent with how the spec handles filesystems currently.

But I have to admit that this might not be the best idea. This is just a workaround to deal with the fact that the spec currently does not have any explicit filesystem handling mechanism. cc @jdef

Reply:
After an offline discussion with @gnufied it seems reasonable to leave this as is. Please drop my comment.